File: migration/src/connectors/gdelt.ts
/**
 * GDELT News API Connector
 *
 * Fetches news from the Global Database of Events, Language, and Tone (GDELT).
 * https://blog.gdeltproject.org/gdelt-doc-2-0-api-debuts/
 */

import type { GdeltResponse, GdeltArticle, NormalizedNewsItem, NewsCategory } from '../types';
import { config, GDELT_QUERIES, containsAlertKeyword, detectRegion, detectTopics } from '../utils/config';
import { generateId, parseGdeltDate, nowISO, delay, logger } from '../utils';

export interface GdeltConnectorOptions {
  /** Number of articles to fetch per category (max 250) */
  maxRecords?: number;
  /** Timespan for articles (e.g., "7d", "24h") */
  timespan?: string;
  /** Language filter */
  language?: string;
  /** Custom CORS proxy URL (optional) */
  corsProxy?: string;
}

const DEFAULT_OPTIONS: Required<GdeltConnectorOptions> = {
  maxRecords: 20,
  timespan: '7d',
  language: 'english',
  corsProxy: '',
};

/**
 * Transform a GDELT article to normalized schema
 */
function transformArticle(
  article: GdeltArticle,
  category: NewsCategory,
  index: number
): NormalizedNewsItem {
  const title = article.title || '';
  const alert = containsAlertKeyword(title);
  const region = detectRegion(title);
  const topics = detectTopics(title);

  return {
    id: generateId(article.url, `gdelt-${category}`),
    source: article.domain || 'GDELT',
    url: article.url,
    title,
    published_at: parseGdeltDate(article.seendate),
    fetched_at: nowISO(),
    authors: [],
    summary: '',
    content_text: '',
    tickers: [],
    topics,
    language: article.language || 'en',
    metadata: {
      category,
      is_alert: alert.is_alert,
      alert_keyword: alert.keyword,
      region: region ?? undefined,
      domain: article.domain,
      image_url: article.socialimage,
      raw: article,
    },
  };
}

/**
 * Build GDELT API URL
 */
function buildGdeltUrl(category: NewsCategory, options: Required<GdeltConnectorOptions>): string {
  const baseQuery = GDELT_QUERIES[category];
  if (!baseQuery) {
    throw new Error(`Unknown category: ${category}`);
  }

  const fullQuery = `${baseQuery} sourcelang:${options.language}`;
  const params = new URLSearchParams({
    query: fullQuery,
    timespan: options.timespan,
    mode: 'artlist',
    maxrecords: options.maxRecords.toString(),
    format: 'json',
    sort: 'date',
  });

  return `${config.gdeltBaseUrl}/api/v2/doc/doc?${params.toString()}`;
}

/**
 * Fetch news from GDELT for a single category
 */
export async function fetchGdeltCategory(
  category: NewsCategory,
  options: GdeltConnectorOptions = {}
): Promise<NormalizedNewsItem[]> {
  const opts = { ...DEFAULT_OPTIONS, ...options };
  const url = buildGdeltUrl(category, opts);

  logger.debug(`Fetching GDELT ${category}:`, url);

  try {
    // Use CORS proxy if specified
    const fetchUrl = opts.corsProxy ? opts.corsProxy + encodeURIComponent(url) : url;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), config.requestTimeout);

    const response = await fetch(fetchUrl, {
      signal: controller.signal,
      headers: { Accept: 'application/json' },
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const contentType = response.headers.get('content-type');
    if (!contentType?.includes('application/json')) {
      logger.warn(`GDELT ${category}: Non-JSON response:`, contentType);
      return [];
    }

    const text = await response.text();
    let data: GdeltResponse;

    try {
      data = JSON.parse(text);
    } catch {
      logger.warn(`GDELT ${category}: Invalid JSON:`, text.slice(0, 100));
      return [];
    }

    if (!data?.articles) {
      logger.debug(`GDELT ${category}: No articles in response`);
      return [];
    }

    logger.info(`GDELT ${category}: Fetched ${data.articles.length} articles`);

    return data.articles.map((article, index) => transformArticle(article, category, index));
  } catch (error) {
    if ((error as Error).name === 'AbortError') {
      logger.error(`GDELT ${category}: Request timeout`);
    } else {
      logger.error(`GDELT ${category}:`, error);
    }
    return [];
  }
}

/**
 * Fetch news from all GDELT categories
 */
export async function fetchAllGdelt(
  categories: NewsCategory[] = ['politics', 'tech', 'finance', 'gov', 'ai', 'intel'],
  options: GdeltConnectorOptions = {}
): Promise<NormalizedNewsItem[]> {
  const allItems: NormalizedNewsItem[] = [];

  for (let i = 0; i < categories.length; i++) {
    const category = categories[i];

    // Add delay between requests to avoid rate limiting
    if (i > 0) {
      await delay(config.delayBetweenRequests);
    }

    const items = await fetchGdeltCategory(category, options);
    allItems.push(...items);
  }

  return allItems;
}

/**
 * GDELT Connector class for stateful usage
 */
export class GdeltConnector {
  private options: Required<GdeltConnectorOptions>;

  constructor(options: GdeltConnectorOptions = {}) {
    this.options = { ...DEFAULT_OPTIONS, ...options };
  }

  async fetchCategory(category: NewsCategory): Promise<NormalizedNewsItem[]> {
    return fetchGdeltCategory(category, this.options);
  }

  async fetchAll(
    categories: NewsCategory[] = ['politics', 'tech', 'finance', 'gov', 'ai', 'intel']
  ): Promise<NormalizedNewsItem[]> {
    return fetchAllGdelt(categories, this.options);
  }
}

File: py_migration/news_scanner/connectors/gdelt.py
"""
GDELT News API Connector

Fetches news from the Global Database of Events, Language, and Tone (GDELT).
https://blog.gdeltproject.org/gdelt-doc-2-0-api-debuts/
"""

import asyncio
from typing import Optional
from urllib.parse import urlencode

import aiohttp

from .base import BaseConnector
from ..models import NormalizedNewsItem, NewsMetadata
from ..utils import config, logger, generate_id, parse_gdelt_date, now_iso
from ..utils.config import GDELT_QUERIES
from ..analytics import TopicDetector, AlertDetector


class GdeltConnector(BaseConnector):
    """
    Connector for GDELT news API.

    Example usage:
        connector = GdeltConnector()

        # Fetch single category
        items = await connector.fetch_category("finance")

        # Fetch multiple categories
        items = await connector.fetch(categories=["finance", "tech"])

        # With options
        items = await connector.fetch(
            categories=["finance"],
            max_records=50,
            timespan="24h"
        )
    """

    def __init__(
        self,
        base_url: Optional[str] = None,
        timeout: int = 15,
        max_records: int = 20,
        timespan: str = "7d",
        language: str = "english",
    ):
        """
        Initialize GDELT connector.

        Args:
            base_url: GDELT API base URL.
            timeout: Request timeout in seconds.
            max_records: Max articles per request (max 250).
            timespan: Time range for articles (e.g., "7d", "24h").
            language: Language filter.
        """
        self.base_url = base_url or config.gdelt_base_url
        self.timeout = timeout
        self.max_records = min(max_records, 250)
        self.timespan = timespan
        self.language = language

        # Analytics (modular)
        self.topic_detector = TopicDetector()
        self.alert_detector = AlertDetector()

    async def fetch(
        self,
        categories: Optional[list[str]] = None,
        **kwargs
    ) -> list[NormalizedNewsItem]:
        """
        Fetch news from all specified categories.

        Args:
            categories: List of categories to fetch. Defaults to all.

        Returns:
            List of normalized news items.
        """
        if categories is None:
            categories = list(GDELT_QUERIES.keys())

        all_items = []

        for i, category in enumerate(categories):
            if i > 0:
                await asyncio.sleep(config.delay_between_requests)

            items = await self.fetch_category(category, **kwargs)
            all_items.extend(items)

        return all_items

    async def fetch_category(
        self,
        category: str,
        max_records: Optional[int] = None,
        timespan: Optional[str] = None,
        **kwargs
    ) -> list[NormalizedNewsItem]:
        """
        Fetch news for a specific category.

        Args:
            category: News category.
            max_records: Override default max records.
            timespan: Override default timespan.

        Returns:
            List of normalized news items.
        """
        query = GDELT_QUERIES.get(category)
        if not query:
            logger.warning(f"Unknown GDELT category: {category}")
            return []

        # Build query
        full_query = f"{query} sourcelang:{self.language}"
        params = {
            "query": full_query,
            "timespan": timespan or self.timespan,
            "mode": "artlist",
            "maxrecords": max_records or self.max_records,
            "format": "json",
            "sort": "date",
        }

        url = f"{self.base_url}/api/v2/doc/doc?{urlencode(params)}"
        logger.debug(f"Fetching GDELT {category}: {url}")

        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(url, timeout=self.timeout) as response:
                    if response.status != 200:
                        logger.warning(f"GDELT {category}: HTTP {response.status}")
                        return []

                    content_type = response.headers.get('content-type', '')
                    if 'json' not in content_type:
                        logger.warning(f"GDELT {category}: Non-JSON response")
                        return []

                    data = await response.json()

            articles = data.get("articles", [])
            logger.info(f"GDELT {category}: {len(articles)} articles")

            return [
                self._transform_article(article, category)
                for article in articles
            ]

        except asyncio.TimeoutError:
            logger.error(f"GDELT {category}: Timeout")
            return []
        except aiohttp.ClientError as e:
            logger.error(f"GDELT {category}: {e}")
            return []
        except Exception as e:
            logger.error(f"GDELT {category}: {e}")
            return []

    def _transform_article(self, article: dict, category: str) -> NormalizedNewsItem:
        """Transform GDELT article to normalized schema."""
        title = article.get("title", "")
        url = article.get("url", "")

        # Use modular analytics
        topics = self.topic_detector.detect(title)
        alert = self.alert_detector.detect(title)

        return NormalizedNewsItem(
            id=generate_id(url, f"gdelt-{category}"),
            source=article.get("domain", "GDELT"),
            url=url,
            title=title,
            published_at=parse_gdelt_date(article.get("seendate", "")),
            fetched_at=now_iso(),
            authors=[],
            summary="",
            content_text="",
            tickers=[],
            topics=topics,
            language=article.get("language", "en"),
            metadata=NewsMetadata(
                category=category,
                is_alert=alert.is_alert,
                alert_keyword=alert.keyword,
                domain=article.get("domain"),
                image_url=article.get("socialimage"),
                raw=article,
            ),
        )
